framework/src/main/java/org/checkerframework/framework/flow/CFAbstractValue.java

```diff
@@ -339,13 +336,14 @@ public MostSpecificVisitor(TypeMirror aTypeMirror, TypeMirror bTypeMirror, V bac
           && gatf.hasQualifierParameterInHierarchy(TypesUtils.getTypeElement(bTypeMirror), top)) {
         // Both types have qualifier parameters, so they are related by invariance rather
         // than subtyping.
-        if (qualHierarchy.isSubtype(a, b) && qualHierarchy.isSubtype(b, a)) {
+        if (qualHierarchy.isSubtypeShallow(a, aTypeMirror, b, bTypeMirror)
+            && qualHierarchy.isSubtypeShallow(b, bTypeMirror, a, aTypeMirror)) {
           return b;
         }
-      } else if (qualHierarchy.isSubtype(a, b)) {
-        return a;
-      } else if (qualHierarchy.isSubtype(b, a)) {
-        return b;
+      } else if (qualHierarchy.isSubtypeShallow(a, aTypeMirror, b, bTypeMirror)) {
+        return qualHierarchy.lowestQualifier(a, b);
```

Comment:
I'm unclear why to call `lowestQualifier` if we already know that `a` is a (shallow) subtype of `b`.  That is, is there something more specific we can do when we have that knowledge?

I see that `lowestQualifier` breaks nondeterminism, but does that nondeterminism (no unique lowest type) indicate a problem in the Checker Framework?  When does it occur?  Documentation here and/or in `lowestQualifier` would help me to understand.